Initial commit - ros2_generate_interface_docs

ahcorde opened the pull request:
This includes the generation of static HTML for actions, messages and services. The idea is to create this kind of pages: http://docs.ros.org/melodic/api/std_srvs/html/srv/Empty.html

As included in the README.md

# ros2_generate_interface_docs

This package will generate the static HTML documentation for messages, services and actions.
It will create the documentation for all the available interfaces in your ROS2 workspace.
It creates a folder with the name of the package (for example: action_msgs).
In this folder it create another folders for actions, messages or services if they exists, inside each folder you will find the correspoding documentation of the interfaces.

For example:

```plain
api/html/action_msgs/
   - msg/
     - GoalInfo.html
     - GoalStatus.html
     - GoalStatus.html
   - srv/
     - CancelGoal.html
api/html/test_msgs/
  - msg/
    - Arrays.html
    - BasicTypes.html
    - BoundedSequences.html
    - Builtins.html
    - Constants.html
    - Defaults.html
    - Empty.html
    - MultiNested.html
    - Nested.html
    - Strings.html
    - UnboundedSequences.html
    - WStrings.html
  - srv/
    - Arrays.html
    - BasicTypes.html
    - Empty.html
  - action
    - Fibonacci.html
    - NestedMessage.html
```

To generate the interfaces: compile the package and execute:

```bash
ros2 run ros2_generate_interface_docs ros2_generate_interface_docs
```

Options:

 - ``--outputdir[default: 'api/']``: Output directory where the html will be placed.

Signed-off-by: ahcorde <ahcorde@gmail.com>

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/main.py:
```diff
@@ -0,0 +1,101 @@
+    args = parser.parse_args(argv)
+
+    output_dir = args.outputdir
+    if not os.path.exists(output_dir):
```
This can just be:
os.makedirs(args.output_dir, exist_ok=True)

Same for line 44-45

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py:
```diff
@@ -0,0 +1,85 @@
+        f.write(msg_style_css)
+
+
+def load_tmpl(filename):
```
Try not to use abbreviations. `load_template(filename)`

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/action_utils.py:
```diff
@@ -0,0 +1,54 @@
+    package, interface, base_type = utils.resource_name(action)
+    d = {'name': action, 'ext': 'action', 'type': 'Action',
+         'package': package, 'base_type': base_type,
+         'date': str(time.strftime('%a, %d %b %Y %H:%M:%S'))}
```
You use this time formatting string a lot, it might be better to create a constant in utils.py that each file can use.

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/action_utils.py:
```diff
@@ -0,0 +1,54 @@
+         'action_list': '',
+         'date': str(time.strftime('%a, %d %b %Y %H:%M:%S'))}
+    if actions:
+        d['action_list'] = """<h2>Action types</h2>
```
What do you think of creating a constant out of this string and moving it to the top of the file, so the method logic is a bit simpler to understand.

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/action_utils.py:
```diff
@@ -0,0 +1,54 @@
+    return action_template % d
+
+
+def generate_action_index(package, file_d, actions):
```
Try to avoid abbreviations where ever possible.

```suggestion
def generate_action_index(package, file_directory, actions):
```

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/main.py:
```diff
@@ -0,0 +1,101 @@
+    msg_template = utils.load_tmpl('msg.template')
+
+    message_interfaces = get_message_interfaces()
+    for package_name in sorted(message_interfaces):
```
Since you make use of the values as well, you could do something like:

```
for package_name, message_names in sorted(message_interfaces.items(), key=lambda item: item[0]):
```

Then the reader has a better understanding of the type of value.

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/main.py:
```diff
@@ -0,0 +1,101 @@
+
+    message_interfaces = get_message_interfaces()
+    for package_name in sorted(message_interfaces):
+        package_d = os.path.join(html_dir, package_name)
```
'_d' is not a common abbreviation that I'm aware of. What if you used '_dir' or just '_directory' for each variable below?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/main.py:
```diff
@@ -0,0 +1,101 @@
+    for package_name in sorted(message_interfaces):
+        package_d = os.path.join(html_dir, package_name)
+        msg_d = os.path.join(package_d, 'msg')
+        if not os.path.exists(msg_d):
```
os.makedirs(msg_dir, exist_ok=True)

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/main.py:
```diff
@@ -0,0 +1,101 @@
+            os.makedirs(msg_d)
+        msg_utils.generate_msg_index(package_name, package_d, message_interfaces[package_name])
+        for message_name in sorted(message_interfaces[package_name]):
+            print(message_name)
```
Is there a more informational message to print out here, or was this left over from debugging?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/main.py:
```diff
@@ -0,0 +1,101 @@
+            text = msg_utils.generate_msg_doc('%s/%s' % (package_name, message_name),
+                                              msg_template,
+                                              package_d + '/' + message_name + '.html')
+            with open(package_d + '/' + message_name + '.html', 'w') as f:
```
Use `os.path.join`

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/main.py:
```diff
@@ -0,0 +1,101 @@
+            with open(package_d + '/' + message_name + '.html', 'w') as f:
+                f.write(text)
+
+    action_template = utils.load_tmpl('action.template')
```
The logic for actions and services looks pretty similar to messages. Is there a way to refactor for better code reuse?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/msg_utils.py:
```diff
@@ -0,0 +1,107 @@
+         'action_list': '',
+         'date': str(time.strftime('%a, %d %b %Y %H:%M:%S'))}
+    if msgs:
+        d['msg_list'] = """<h2>Message types</h2>
```
Move string to a constant at the top of the file?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/srv_utils.py:
```diff
@@ -0,0 +1,123 @@
+    return OrderedDict(zip([s[1:] for s in srv.__slots__], srv.SLOT_TYPES))
+
+
+def _generate_srv_text_from_spec(package, type_, spec, buff=None, indent=0):
```
This duplicates a lot of functionality already written in msg_utils.py. Is there a better way to reuse the same code?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def _href(link, text):
+    return '<a href="%(link)s">%(text)s</a>' % locals()
```
The use of locals() here seems like it would create some hard to debug behavior. Could these values just be passed in as a '**kwargs'?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py:
```diff
@@ -0,0 +1,85 @@
+
+def generate_raw_text(raw_text):
+    s = ''
+    for line in raw_text.split('\n'):
```
raw_text.split(os.linesep)

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py:
```diff
@@ -0,0 +1,85 @@
+
+def resource_name(resource):
+    """Return the resource name."""
+    if '/' in resource:
```
The nested logic in this function seems harder to follow than it should be:

If you used early returns, it might be simpler to visually parse
```
if '/' not in resource:
  return '', '', resource
values = resource.split('/')
if len(values) != 3:
    raise ValueError(...)
return tuple(values)
``` 

Also, try not to use abbreviations. `val` -> `values`

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py:
```diff
@@ -0,0 +1,85 @@
+def copy_css_style(folder_name):
+    """Copy the css style file in the folder_name."""
+    style_css = load_tmpl('styles.css')
+    with open(folder_name + '/styles.css', 'w') as f:
```
Use os.path.join(). Same comment for line 63.

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py:
```diff
@@ -0,0 +1,85 @@
+    :returns: cached file contents
+    """
+    filename = os.path.join(get_templates_dir(), filename)
+    if not os.path.isfile(filename):
```
Is this check necessary? Line 80 will already raise an exception if it doesn't exist.

ahcorde replied:
if the file doesn't exists then we will close this "cleanly", otherwise a exception will arraise.

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py:
```diff
@@ -0,0 +1,85 @@
+    with open(filename, 'r') as f:
+        content = f.read()
+        if not content:
+            sys.stderr.write("Template file '%s' is empty\n" % (filename))
```
Use 'raise Exception' instead

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def generate_raw_text(raw_text):
+    s = ''
```
Try not to use single letter variable names. What's a more descriptive name for this?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/msg_utils.py:
```diff
@@ -0,0 +1,107 @@
+    message_fields = imported_message.get_fields_and_field_types()
+    for field in message_fields.keys():
+        if(message_fields[field] not in BASIC_TYPES and
+           'string' not in message_fields[field]):
```
Using parens here to break this line results in an indentation for line 55 that aligns it with line 56. Maybe a `\` would be better so you can tab indent line 55 one further?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/msg_utils.py:
```diff
@@ -0,0 +1,107 @@
+        if(message_fields[field] not in BASIC_TYPES and
+           'string' not in message_fields[field]):
+            slot_type = get_message_slot_types(imported_message)[field]
+            if(isinstance(slot_type, AbstractNestedType)):
```
Lines 57 - 73 are the meat of this function, but they are pretty tough to follow and a lot gets reused. Since each buff.write takes the same string formatter, I would suggest creating three variables that get created individually, instead of trying to do it all inside buff.write().

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/msg_utils.py:
```diff
@@ -0,0 +1,107 @@
+           'string' not in message_fields[field]):
+            slot_type = get_message_slot_types(imported_message)[field]
+            if(isinstance(slot_type, AbstractNestedType)):
+                if(isinstance(slot_type.value_type, NamespacedType)):
```
Is there supposed to be an else for this if statement?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/msg_utils.py:
```diff
@@ -0,0 +1,107 @@
+                                                  field))
+            else:
+                pkg_name, msg_name = message_fields[field].split('/')
+                buff.write('%s%s %s<br />' % ('nbsp;' * indent,
```
If you create an `indentation` variable this will likely get addressed, but I think 'nbsp' is missing a `&` here.

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/action_utils.py:
```diff
@@ -0,0 +1,54 @@
+
+
+def generate_action_index(package, file_d, actions):
+    d = {'package': package, 'msg_list': '', 'srv_list': '',
```
Please don't use single letter variable names.

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/action_utils.py:
```diff
@@ -0,0 +1,54 @@
+  <ul>
+%s
+  </ul>
+</div>""" % '\n'.join([' <li>%s</li>' % utils._href('../../html/' + package + '/' + m + '.html', m)
```
Please don't use single letter variable names. What is m supposed to be?

brawner commented on ros2_generate_interface_docs/ros2_generate_interface_docs/action_utils.py:
```diff
@@ -0,0 +1,54 @@
+</div>""" % '\n'.join([' <li>%s</li>' % utils._href('../../html/' + package + '/' + m + '.html', m)
+                       for m in actions])
+    msg_index_template = utils.load_tmpl('msg-index.template')
+    file_p = os.path.join(file_d, 'index-msg.html')
```
I see names like this a lot in this code. It would be better to just use full variable names instead of abbreviations.

ahcorde commented:
Created the ticket https://github.com/ros2/ros2_generate_interface_docs/issues/2 *Feature to allow for generating docs for a single package*.

This PR will add the basic functionality

Blast545 commented on README.md:
```diff
@@ -1 +1,49 @@
+This package will generate the static HTML documentation for messages, services and actions.
+It will create the documentation for all the available interfaces in your ROS2 workspace.
+It creates a folder with the name of the package (for example: action_msgs).
+In this folder it create another folders for actions, messages or services if they exists, inside each folder you will find the correspoding documentation of the interfaces.
```
```suggestion
In this folder it creates other directories for actions, messages or services if they exist, inside each folder you will find the corresponding documentation of the interfaces.
```

Blast545 commented on ros2_generate_interface_docs/setup.py:
```diff
@@ -0,0 +1,39 @@
+        'Topic :: Software Development',
+    ],
+    description='Tool to generate interfaces public API documentation',
+    license='Apache License, Version 2.0, BSD',
```
Is it intended to mention BSD?

Blast545 commented on ros2_generate_interface_docs/test/test_copyright.py:
```diff
@@ -0,0 +1,24 @@
+@pytest.mark.copyright
+@pytest.mark.linter
+def test_copyright():
+    excluded = ['cpplint.py']
```
May I ask, is this file part of the package? I see its skipped in the added linters but I don't see anywhere

dirk-thomas replied:
Same in the other linter tests.

dirk-thomas commented on ros2_generate_interface_docs/setup.py:
```diff
@@ -0,0 +1,39 @@
+
+setup(
+    name=package_name,
+    version='0.0.1',
```
Before the first release this should be 0.0.0 since 0.0.1 hasn't been released yet.

dirk-thomas commented on ros2_generate_interface_docs/ros2_generate_interface_docs/action_utils.py:
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python3
```
Why a shebang line?

Same for multiple files below.

dirk-thomas commented on README.md:
```diff
@@ -1 +1,49 @@
+# ros2_generate_interface_docs
+
+This package will generate the static HTML documentation for messages, services and actions.
+It will create the documentation for all the available interfaces in your ROS2 workspace.
```
Why should this be tied to a ROS workspace at all and not take arguments pointing to e.g. packages?

dirk-thomas replied:
Also how does it handle if interface files are being generated at build time by a package?

ahcorde replied:
I created a ticket for your first comment https://github.com/ros2/ros2_generate_interface_docs/issues/2

can you give some pointers to a interface generated at build time ?

dirk-thomas commented on ros2_generate_interface_docs/ros2_generate_interface_docs/msg_utils.py:
```diff
@@ -0,0 +1,75 @@
+    with open(file_path, 'r', encoding='utf-8') as h:
+        spec = h.read().rstrip()
+
+    d['fancy_text'] = _generate_msg_text_from_spec(package, base_type, spec)
```
Please use more descriptive names. From reading this it is unclear what about this text is "fancy".

dirk-thomas commented on ros2_generate_interface_docs/ros2_generate_interface_docs/msg_utils.py:
```diff
@@ -0,0 +1,75 @@
+    package_message_dic['date'] = str(time.strftime('%a, %d %b %Y %H:%M:%S'))
+    if msgs:
+        package_message_dic['msg_list'] = message_list_str % '\n'.join(
+            [' <li>%s</li>' % utils._href('../../html/' + package + '/' + msg + '.html', msg)
```
From an architecture point of view I wouldn't expect this logic to be embedded in Python modules but in the template instead - assuming the template is changed into an `.em` template. Same for the formatting of the date. Basically separating the view from the model.

Same in various other places.

dirk-thomas commented on ros2_generate_interface_docs/ros2_generate_interface_docs/srv_utils.py:
```diff
@@ -0,0 +1,91 @@
+from rosidl_runtime_py.import_message import import_message_from_namespaced_type
+
+try:
+    from cStringIO import StringIO
```
This is Python 2 specific logic which doesn't seem necessary.

ahcorde commented:
Splitted in 4 PRs:

> - package stubs and miscellaneous file

*Added the basic structure for the package* https://github.com/ros2/ros2_generate_interface_docs/pull/3

> - logic for e.g. messages
>     - this will need to decide how the package should be used - which could be a separate PR / ticket too
 - Added basic logic to generate msgs.
 - Moved some logic to .em files

 > - logic for services
 > - logic for actions

I will update services and actions when we discuss the basics for messages

I don't have write permissions I can't build PR on top of other PR. Waiting for writing permissions.

ahcorde commented:
Closing this PR in favour of these other PRs
 - Added the basic structure for the package https://github.com/ros2/ros2_generate_interface_docs/pull/3
 - Added logic to generate msg static HTML docs: https://github.com/ros2/ros2_generate_interface_docs/pull/4
